### PR TITLE
Atlas 31 ancestors in concept tab

### PR DESCRIPTION
**Pull request by agackovka**
https://github.com/OHDSI/Atlas/issues/31

### CODE REVIEW

File: js/components/cohort-definitions/cohort-definition-manager.html [hunk 1]
```diff
@@ -565,3 +565,46 @@
 		</div>
 	</div>
 </div>
+<div class="modal" id="ancestorsModal" tabindex="-1" role="dialog">
```
Comment (pavgra) [inline js/components/cohort-definitions/cohort-definition-manager.html]: Move to separate component

File: js/components/cohort-definitions/cohort-definition-manager.html [hunk 2]
```diff
@@ -565,3 +565,46 @@
...
+	<div class="modal-dialog" role="document">
+		<div class="modal-content">
+			<div class="modal-body">
+				<div>
+					<div class="paddedWrapper">
+						<table class="conceptSetTable stripe compact hover" cellspacing="0" width="100%" data-bind="dataTable:{
+							data:ancestors,
+							options: {
+									dom: 'Clfiprt',
```
Comment (pavgra) [inline js/components/cohort-definitions/cohort-definition-manager.html]: What is this?

File: js/components/cohort-definitions/cohort-definition-manager.js [hunk 3]
```diff
@@ -102,6 +102,41 @@ define(['knockout', 'text!./cohort-definition-manager.html',
...
+		
+		self.ancestors = ko.observableArray();
+		
+		self.showAncestorsModal = function(conceptId) {
+			self.ancestors(self.model.includedConcepts()
+				.find(v => v.CONCEPT_ID === conceptId)
+				.ANCESTORS
+				.map(v => ({concept: v})));
+			if (!_.isEmpty(self.ancestors())) {
+				$('#ancestorsModal').modal();
```
Comment (pavgra) [inline js/components/cohort-definitions/cohort-definition-manager.js]: Use bindings

File: js/components/cohort-definitions/cohort-definition-manager.js [hunk 4]
```diff
@@ -102,6 +102,41 @@ define(['knockout', 'text!./cohort-definition-manager.html',
...
+				const rows = api.rows({page: 'current'});
+				const data = rows.data();
+				const promise = self.model.loadAndApplyAncestors(data);
+				const columnIndex = self.includedConceptsColumns.findIndex(v => v.data === 'ANCESTORS');
+				promise.then(() => {
+					api.cells(null, columnIndex).invalidate();
+					rows.nodes().each((element, index) => {
+						const rowData = data[index];
+						self.model.contextSensitiveLinkColor(element, rowData);
+						$(element).children(`td:nth-child(${columnIndex})`).children(':first')
```
Comment (pavgra) [inline js/components/cohort-definitions/cohort-definition-manager.js]: We haven't discussed such dirty solution

File: js/components/conceptset/conceptset-manager.html [hunk 5]
```diff
@@ -383,3 +383,47 @@
     </div>
   </div>
 </div>
+
+<div class="modal" id="ancestorsModal" tabindex="-1" role="dialog">
```
Comment (pavgra) [inline js/components/conceptset/conceptset-manager.html]: Reuse the component, which should have been created above

File: js/components/conceptset/conceptset-manager.js [hunk 6]
```diff
@@ -369,8 +372,34 @@ define(['knockout',
 		}, {
 			title: 'Vocabulary',
 			data: 'VOCABULARY_ID'
+		}, {
+			title: 'Ancestor',
+			data: 'ANCESTORS',
+			render: (s, p, d) => '<a class="clickable">' + d.ANCESTORS.length + '</a>'
 		}];
 
+		self.includedDrawCallback = function (settings) {
```
Comment (pavgra) [inline js/components/conceptset/conceptset-manager.js]: Remove copy-paste

File: js/components/conceptset/conceptset-manager.js [hunk 7]
```diff
@@ -992,7 +1023,17 @@ define(['knockout',
 					includeMapped: true
 				})
 			});
-
+		
+		self.showAncestorsModal = function(conceptId) {
```
Comment (pavgra) [inline js/components/conceptset/conceptset-manager.js]: Remove copy-paste

File: js/main.js [hunk 8]
```diff
@@ -272,30 +272,74 @@ requirejs(['bootstrap'], function () { // bootstrap must come first
...
 		});
 
-		pageModel.loadIncluded = function () {
-			pageModel.loadingIncluded(true);
-			var includedPromise = $.Deferred();
-
-			$.ajax({
-				url: sharedState.vocabularyUrl() + 'lookup/identifiers',
+		pageModel.loadAncestors = function(ancestors, descendants) {
+			return $.ajax({
```
Comment (pavgra) [inline js/main.js]: Move such things to services (either find existing in WebAPIProvider or create a new one)

File: js/components/conceptset/conceptset-manager.js [hunk 9]
```diff
@@ -921,6 +931,8 @@ define(['knockout',
 			}
 		}
 
+		self.title = 'title <br> titl2 second line';
```
Comment (pavgra) [inline js/components/conceptset/conceptset-manager.js]: what is this?

**Comment by anthonysena** [comment]
Thanks @agackovka for making these changes here and to OHDSI/WebAPI#369. After another review, I found another issue. Here is the flow:

Define a concept set that includes an excluded concept. For example, here is one I was using for testing:

`{"items" :[{"concept":{"CONCEPT_ID":443392,"CONCEPT_NAME":"Malignant neoplastic disease","STANDARD_CONCEPT":"S","INVALID_REASON":"V","CONCEPT_CODE":"363346000","DOMAIN_ID":"Condition","VOCABULARY_ID":"SNOMED","CONCEPT_CLASS_ID":"Clinical Finding","INVALID_REASON_CAPTION":"Valid","STANDARD_CONCEPT_CAPTION":"Standard"},"isExcluded":false,"includeDescendants":true,"includeMapped":false},{"concept":{"CONCEPT_ID":4177112,"CONCEPT_NAME":"Malignant tumor of trachea","STANDARD_CONCEPT":"S","INVALID_REASON":"V","CONCEPT_CODE":"363432004","DOMAIN_ID":"Condition","VOCABULARY_ID":"SNOMED","CONCEPT_CLASS_ID":"Clinical Finding","INVALID_REASON_CAPTION":"Valid","STANDARD_CONCEPT_CAPTION":"Standard"},"isExcluded":true,"includeDescendants":false,"includeMapped":false}]}`

When navigating to the included concepts, I found that I was receiving the following error:

image

Tracing through the code, I found that to compute the ancestors, you are passing in the concepts that are defined in the concept set (regardless of the use of the 'exclude' flag) to WebAPI to find common ancestors. Then in Atlas/js/main.js: https://github.com/OHDSI/Atlas/pull/614/files#diff-77ab163850ef0fa98d6a9ee9fb6bbcafR304 attempts to creates a collection of ancestors based on the concepts included in the concept set. What happens is that since you are attempting to match the ancestors up to the `pageModel.includedConceptsMap` you wind up with an undefined ancestor since the concept we specified as 'excluded' is not in the map. That then causes the error above to happen.

The other thing I noticed is that the call to /lookup/identifiers/ancestors is called twice? I didn't dive into what is happening there but thought I'd bring this to your attention as well. Thanks!

**Comment by agackovka** [comment]
@anthonysena thanks for your review, the mentioned bug has been fixed.
Regarding two xhrs: calculation of ancestors for whole table is quite expensive task, so initially we calculate ancestors for first page, and simultaneously in background we calculate whole table ancestors, because we should be able to sort by 'Ancestor' column

**Comment by anthonysena** [comment]
@agackovka thanks for making the changes. I think we're looking good on the Atlas side but I did leave a comment on the cooresponding OHDSI/WebAPI#369 pull request related to parameter limits. Thanks!

**Pull request merged**